consensus/src/sync/live/block_queue/mod.rs

```diff
@@ -304,12 +301,78 @@ impl<N: Network> BlockQueue<N> {
+    ///
+    /// At most two block per peer per block height will be buffered leaving enough room for a bock and the
+    /// skip block making it obsolete. For benign peers that is sufficient, as they will never propagate
+    /// more than that.
+    ///
+    /// ## Returns
+    /// * `true` if the block was added to the buffer
+    /// * `false` otherwise
     fn insert_block_into_buffer(&mut self, block: Block, pubsub_id: Option<N::PubsubId>) -> bool {
-        self.buffer
-            .entry(block.block_number())
-            .or_default()
-            .insert(block.hash(), (block, pubsub_id))
-            .is_some()
+        let peer_id = match pubsub_id {
+            // Blocks without a pubsub_id are requested and will always be buffered.
+            None => {
+                let map = self.buffer.entry(block.block_number()).or_default();
+                if map.contains_key(&block.hash()) {
+                    return false;
+                }
+                assert!(map.insert(block.hash(), (block, None)).is_none());
+                return true;
+            }
+            // For blocks with a pubsub id the propagation source is the peer which can
+            // only have 2 items buffered.
+            Some(ref pubsub_id) => pubsub_id.propagation_source(),
+        };
+
+        // Get the entry for the block number if it exists.
+        // Otherwise add the entry and add the block to it and return.
+        let map = match self.buffer.entry(block.block_number()) {
+            BTreeMapEntry::Occupied(occupied_entry) => occupied_entry.into_mut(),
+            BTreeMapEntry::Vacant(vacant_entry) => {
+                // Trivially okay to buffer as nothing is buffered yet.
+                vacant_entry.insert(HashMap::from([(block.hash(), (block, pubsub_id))]));
+                return true;
+            }
+        };
+
+        let mut blocks_by_peer = 0;
+        for (hash, (_block, pubsub)) in map.iter() {
+            if *hash == block.hash() {
+                // Block is already buffered. Ignore it here.
+                return false;
+            }
+
+            // Extract the propagation source or move on to the next item
+            let propagation_source = match pubsub {
+                Some(pubsub_id) => pubsub_id.propagation_source(),
+                None => continue,
+            };
+
+            // Check if the source is the current peer
+            if propagation_source == peer_id {
+                // Increase the counter
+                blocks_by_peer += 1;
+                // Check if that is the count exceeds the maximum buffer count
+                if blocks_by_peer == Self::MAX_BUFFERED_PER_PEER_PER_HEIGHT {
+                    log::debug!(
+                        "Peer ({:?}) buffer for height {} exceeded limit, dropping block {}",
+                        peer_id,
+                        block.block_number(),
+                        block,
```

Comment:
I'd suggest to use tracing style:
```suggestion
                        "Peer ({:?}) buffer for height {} exceeded limit, dropping block {}",
                        peer_id,
                        block.block_number(),
                        block,
```
```suggestion
                        ?peer_id, block_number=block.block_number(), dropped_block=block
                        "Peer buffer exceeded limit, dropping block",
```